improve autoload_paths logic

meinac opened the pull request:
Normally in rails if you want to add a directory to autoload_paths which is under of the lib directory you have to add whole lib directory to autoload_paths. With this commit we can add only necessary directory into the autoload_paths.
In current behaviour if you add the directory into the autoload_paths like below;

``` ruby
  config.autoload_paths << "#{config.root}/lib/foo"
```

rails can't load the files under the foo directory but now It can.

meinac committed:
improve autoload_paths logic

test cases for new logic

improve tests

repinel commented:
Could `activesupport/test/dependencies_lib/foo/bar.rb` be in `activesupport/test/dependencies`? Otherwise it looks good.

meinac commented:
@repinel hey! Thank you for your advice, I've changed test cases and moved the dependency into the dependencies folder. I hope someone from rails team will review this.

rafaelfranca commented:
Are you sure that doesn't work? I always did that in my applications. Do you have an example where it doesn't work?

meinac commented:
Hi @rafaelfranca , I've created an [example application](https://github.com/meinac/rails_autoload_paths) to show the problem.
Try to run `Foo::Bar.greeting` inside of the rails console, it will raise error. As you can see you can't add a nested directory into the `autoload_paths`. But if you change the addition of new dependency into the `autoload_paths` in `application.rb` like `config.autoload_paths << "#{config.root}/lib"` it works. By this pr we can now add any directory into the `autoload_paths`.

kaspth commented:
When you add a directory to the autoload path, Rails will look up files relative to that directory. You're seeing failures because you're over qualifying the path to Bar.

You wouldn't get at a controller constant with `Controllers::DocumentsController` either. Try referencing just Bar instead of Foo::Bar and it should work :smile:

I'm closing because as I understand it there's no bug in Rails.

rafaelfranca commented:
Right! Now I got it. Exactly as @kaspth explained. I used nested folder but the constant were not nested.

meinac commented:
@kaspth , @rafaelfranca If you try to reach the method like `Bar.greeting` then it will raise this error;
`LoadError: Unable to autoload constant Bar, expected .../autoload_test/lib/foo/bar.rb to define it`

meinac commented:
But if I change the definition of `bar.rb` like;
```ruby
class Bar
...
end
```
it works. 
You are right even if it's in a nested directory the ruby class should not be under a module if I want to add the directory into the `autoload_paths`. But, I think we have to document this.

byroot commented on activesupport/lib/active_support/dependencies.rb:
```diff
@@ -406,6 +406,11 @@ def search_for_file(path_suffix)
         path = File.join(root, path_suffix)
         return path if File.file? path
+        module_path = path_suffix.split('/').first
+        if module_path && root.to_s =~ /#{module_path}$/
```
Wouldn't `root.to_s.end_with?(module_path)` be simpler / nicer ?

byroot commented on activesupport/lib/active_support/dependencies.rb:
```diff
@@ -416,6 +421,10 @@ def search_for_file(path_suffix)
     def autoloadable_module?(path_suffix)
       autoload_paths.each do |load_path|
         return load_path if File.directory? File.join(load_path, path_suffix)
+        if load_path.to_s =~ /#{path_suffix}$/
```
Same here

kaspth commented:
@meinac sure, if you can find a place to improve the autoloading documentation we'd appreciate it. We have an autoloading guide now maybe that's the place to start. Thanks :smile:
